Approved/dyn-type.md

```diff
@@ -0,0 +1,196 @@
+
+
+### Keep the same AST or IR representation for dyn and non-dyn
+
+While conceptually, all global callables effectively become implicitly type parameterized to permit that any argument can be either of non-dyn or dyn type, there is no need to "monomorphize" with respect to this "axis" of the type system; we do not see much value in either a different AST or a different IR representation for dyn vs non-dyn values. This reflects the design goal of alignment with the llvm compilation and optimization logic.
+
+### AST representation needed at validation time
+
+The only place where such a representation is needed is to run the validation (analyzers that run after the initial compilation has been built); what is needed specifically is an AST representation for a specific callable that contains the proper type distinction between dyn and non-dyn types for the purpose of properly propagating that information throughout the implementation such that the analyzers can make use of it. This AST representation can, however, be discarded after the analysis passes for this callable have completed, and is not needed for more than one callable at a time (we do a bottom-up inference). 
+
+### Analyzer pass on each callable parameter
+
+The analyzers are run for one callable at a time (same as it is the case today). They will be run once in (more or less) the same way as they are today, and produce the same attribute they are today - i.e. the first pass of the analyzers analyzes the requirements of the callable under the assumption that all arguments are non-dyn values.
+
+Additionally, we will run the analyzers once for each parameter of the callable. This pass will produce an additional attribute per parameter that indicates the required runtime capability if a specific parameter is `dyn` instead of non-`dyn`. We hence end up with n+1 attribute on the callable, where n is the number of parameters. For each call to the callable, the required capability for that call is given by the sum of the required capabilities for the callable itself plus those for of the `dyn` arguments.
+
+The n+1 attributes need to be manually declared for intrinsic callables (same as it is the case today, just more attributes).
+
+## Approximations
+
+For the purpose of type inference (i.e. how `dyn` is propagated), we make the below approximations.
+
+### Operations always return a dyn
+
+The return type of an operation is always `dyn`. While it is conceivable to write an operation that has a non-`dyn` return value, it is also always possible - and arguably best practice - to split out any computation of a non-`dyn` return value into a function instead.
```

Comment:
Could we raise a warning or a lint message when an operation returns `dyn` only because of this rule? How will we document this approximation?